src/engine/init.c

```diff
@@ -783,6 +783,7 @@ server_init(int argc, char *argv[])
 static void
 server_fini(bool force)
 {
+	crt_rank_abort_all(NULL);
```

Comment:
What problem is being addressed here? Isn't this abort call racing with all other xstreams, who may send new requests after the abort call returns?

Reply:
oh, some RPC completion callback might need those modules, for example IV, so it needs to abort all inflight RPC and execute all completion callback,  then cleanup the module.  Here is what I met in my local test, clearly this IV rpc callback happened after all modules have been cleanup.

[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `/home/wangdi/daos_m_20190301/daos/install/bin/daos_engine -t 8 -x 0 -g daos_ser'.
Program terminated with signal 11, Segmentation fault.
#0  0x0000000000436294 in ivc_on_put (ivns=<optimized out>, iv_value=0x13c1aa00, priv=0x232f5d80) at src/engine/server_iv.c:618
618		rc = entry->iv_class->iv_class_ops->ivc_ent_put(entry,
Missing separate debuginfos, use: debuginfo-install glibc-2.17-323.el7_9.x86_64 hwloc-libs-1.11.8-4.el7.x86_64 libaio-0.3.109-13.el7.x86_64 libatomic-4.8.5-44.el7.x86_64 libgcc-4.8.5-44.el7.x86_64 libibverbs-22.4-5.el7.x86_64 libnl3-3.2.28-4.el7.x86_64 librdmacm-22.4-5.el7.x86_64 libtool-ltdl-2.4.6-29.01.el7.x86_64 libunwind-1.2-2.el7.x86_64 libuuid-2.23.2-65.el7_9.1.x86_64 libyaml-0.1.4-11.el7_0.x86_64 lz4-1.8.3-1.el7.x86_64 numactl-libs-2.0.12-5.el7.x86_64
(gdb) bt
#0  0x0000000000436294 in ivc_on_put (ivns=<optimized out>, iv_value=0x13c1aa00, priv=0x232f5d80) at src/engine/server_iv.c:618
#1  0x00007fb5e7e8aab1 in handle_ivupdate_response (cb_info=0x2b671e0) at src/cart/crt_iv.c:2541
#2  handle_response_internal (arg=0x2b671e0) at src/cart/crt_iv.c:2714
#3  0x00007fb5e7e23f24 in crt_rpc_complete (rpc_priv=rpc_priv@entry=0x35df600, rc=rc@entry=-1018) at src/cart/crt_context.c:351
#4  0x00007fb5e7e2cb3f in crt_ctx_epi_abort (rlink=<optimized out>, arg=<optimized out>) at src/cart/crt_context.c:413
#5  0x00007fb5e8111dc2 in d_hash_table_traverse (htable=htable@entry=0x2b6d9c8, cb=cb@entry=0x7fb5e7e2c920 <crt_ctx_epi_abort>, arg=arg@entry=0x2b673b4) at src/gurt/hash.c:900
#6  0x00007fb5e7e2e99b in crt_context_destroy (crt_ctx=0x2b6d940, force=force@entry=1) at src/cart/crt_context.c:499
#7  0x000000000043c7f6 in dss_srv_handler (arg=0x2a78ad0) at src/engine/srv.c:502
#8  0x00007fb5e70c5daa in ABTD_ythread_func_wrapper () from /home/wangdi/daos_m_20190301/daos/install/bin/../prereq/dev/argobots/lib/libabt.so.1
#9  0x00007fb5e70c5f51 in make_fcontext () from /home/wangdi/daos_m_20190301/daos/install/bin/../prereq/dev/argobots/lib/libabt.so.1
#10 0x0000000000000000 in ?? ()

Reply:
When this segfault happened, server_fini must be calling dss_srv_fini, which was waiting for the xstreams to terminate; server_fini must have not called ds_iv_fini or dss_module_fini yet. So, the class object should be in good state; the segfault might be due to "priv". Any idea how "priv" could point to freed memory? Due to ds_pool_stop -> ds_iv_ns_stop (I don't know)?

Reply:
yes, module_cleanup ->ds_pool_stop->ds_iv_ns_stop->iv_entry_free() will free all IV entries.  

So it is better to abort all RPC before that.

Reply:
After the primary xstream completes the abort call, the system xstream may begin new IV operations, no? (More generally, any xstream may send new RPCs, no?)

Reply:
also something to keep in mind; just because you issued abort to all rpcs, doesnt mean that completion callbacks will be executed before the call returns. This is a non-blocking call, so a completion can happen at much later time. In theory reference counts on various structs that we use should protect against segfaults like listed above; even if the structure is marked as free it should remain 'in use' as long as there is rpc that is still referencing those data

Reply:
all IV is on stream 0, if we abort all RPC on xstream0, then there should no IV RPC & complete callback before module cleanup,  as long as it does not yield.

@frostedcmos  I saw it wait all rpc & complete callback in crt_ctx_epi_abort() ? No? TBH, I think abort should be synchronized, at least there should be a synchronized version abort.

Reply:
ah yes, this version does have sync to wait for all completions; i was thinking of a lower level abort which is just fire-and-forget type of command

Reply:
> all IV is on stream 0, if we abort all RPC on xstream0, then there should no IV RPC & complete callback before module cleanup, as long as it does not yield.

server_fini is called on the primary xstream, not the system xstream.

sm_cleanup implementations are not prohibited from yielding. E.g., does ds_rebuild_leader_stop_all yield? In the future, it's hard to say if the rebuild, obj, and cont sm_cleanup might want to wait for certain resources to be released.

Reply:
ds_rebuild_leader_stop_all() will yield, but after it set sp_stopping, which will blocking all IV RPC.

Hmm, ideally cart_abort_rank actually should abort & stop further rpc send/receive on all context, then it should be safe to start cleanup process.


Let me remove this from the patch, and land those leak fix first.